plugins/teststeps/awsfileupload/awsFileUpload.go

```diff
@@ -0,0 +1,299 @@
+	// Validate the parameter
+	if err := ts.validateAndPopulate(params); err != nil {
+		return nil, err
+	}
+	f := func(ctx xcontext.Context, target *target.Target) error {
+		// expand args
+		path, err := ts.path.Expand(target)
+		if err != nil {
+			return fmt.Errorf("failed to expand argument '%s': %v", ts.path, err)
+		}
+		filename, err := ts.filename.Expand(target)
+		if err != nil {
+			return fmt.Errorf("failed to expand argument dir '%s': %v", ts.filename, err)
+		}
+		var url string
+		// Compress if compress parameter is true
+		if ts.compress {
+			// Create buffer for the compressed data
+			var buf bytes.Buffer
+			// Create the archive and write the output to the "out" Writer
+			buf, err = createTarArchive(path, buf)
+			if err != nil {
+				return fmt.Errorf("error creating an archive: %v", err)
+			}
+			fmt.Println("Tar archive created successfully")
+			filename = filename + ".tar.gz"
+			// Upload file
+			url, err = ts.upload(filename, buf.Bytes())
+			if err != nil {
+				return fmt.Errorf("could not upload the file: %v", err)
+			}
+		} else {
+			// Read the file that should be uploaded
+			bodyBytes, err := ioutil.ReadFile(path)
+			if err != nil {
+				return fmt.Errorf("could not read the file: %v", err)
+			}
+			// Upload file
+			url, err = ts.upload(filename, bodyBytes)
+			if err != nil {
+				return fmt.Errorf("could not upload the file: %v", err)
+			}
+		}
+		// Emit URLtoReport event to get the url into the report
+		if ts.emitURLtoReport {
+			if err := emitEvent(ctx, EventURLtoReport, eventURLtoReportPayload{Msg: url}, target, ev); err != nil {
+				return fmt.Errorf("failed to emit event: %v", err)
+			}
+		}
+		return nil
+	}
+	return teststeps.ForEachTarget(Name, ctx, ch, f)
+}
+
+// Retrieve all the parameters defines through the jobDesc
+func (ts *FileUpload) validateAndPopulate(params test.TestStepParameters) error {
+	// Retrieving parameter as json Raw.Message
+	// validate path and filename
+	ts.path = params.GetOne("path")
+	if ts.path.IsEmpty() {
+		return fmt.Errorf("invalid or missing 'path' parameter, must be exactly one string")
+	}
+	ts.filename = params.GetOne("filename")
+	if ts.filename.IsEmpty() {
+		return fmt.Errorf("invalid or missing 'path' parameter, must be exactly one string")
+	}
+
+	// Retrieving parameter as string
+	// validate s3region
+	param := params.GetOne("s3region")
+	if param.IsEmpty() {
+		return fmt.Errorf("missing 's3region' parameter, must be exactly one string")
+	}
+	ts.s3region = param.String()
+	// validate s3bucket
+	param = params.GetOne("s3bucket")
+	if param.IsEmpty() {
+		return fmt.Errorf("missing 's3bucket' parameter, must be exactly one string")
+	}
+	ts.s3bucket = param.String()
+	// validate s3path
+	param = params.GetOne("s3path")
+	if param.IsEmpty() {
+		return fmt.Errorf("missing 's3path' parameter, must be exactly one string")
+	}
+	ts.s3path = param.String()
+	// retrieve s3credfile
+	ts.s3credfile = params.GetOne("s3credfile").String()
```

Comment:
If s3credfile and s3credprofile are empty, will the plugin fail or work?

Reply:
If they are empty, the session will be open with the default aws credential file path and the default credential profile. But this default also have to be created in order to authenticate to an aws session. So the answer is if this credential file misses, the plugin will always fail. If it is stored as default, the plugin will work, even if the strings are empty.